core/include/jiminy/core/stepper/AbstractStepper.h

```diff
@@ -0,0 +1,82 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+///
+/// \brief      Generic interface for a stepper to integrate the equation of dynamics.
+///
+/// \details    The generalized position of a mechanical system evolves not in a vector space,
+///             but in a Lie group: this is visible for instance in the fact that quaternions
+///             have four components, but only three are meaningful, as quaternion must remain
+///             unitary to represent a rotation. As such, the velocity vector v is not the
+///             term-by-term derivative of the configuration q.
+///             This means that classical stepper implementations cannot be used as is: this
+///             class thus defines an interface for implementing variable-step numerical
+///             solvers on this type of system.
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef JIMINY_ABSTRACT_STEPPER_H
+#define JIMINY_ABSTRACT_STEPPER_H
+
+#include <functional>
+
+#include "jiminy/core/Types.h"
+
+namespace jiminy
+{
+    class Robot;
+    struct systemState_t;
+
+    template<typename F1>
+    class AbstractStepper
+    {
+        public:
+            /// \brief Constructor
+            /// \param[in] f      Dynamics function, with signature a = f(t, q, v)
+            /// \param[in] robots Robots whose dynamics the stepper will work on.
+            AbstractStepper(F1 & f,
+                            std::vector<Robot const *> robots);
+
+            /// \brief Attempt to integrate the system from time t to time t + dt.
+            /// \details Each stepper is responsible for implementing its own error checking routine to determine
+            ///          if an integration step was successful or not. On sucess, the value of the state is
+            ///          updated in place. Regardless of success or failure, the parameter dt can be updated
+            ///          by variable step schemes to indicate the next advised dt.
+            ///
+            /// \param[in]      t     Integration start time.
+            /// \param[in, out] q     System starting position.
+            /// \param[in, out] v     System starting velocity.
+            /// \param[in, out] a     System starting acceleration.
+            /// \param[in, out] dt    Input: desired integration duration. Output: recommanded step size for
+            ///                       variable-step schemes. Constant-step schemes leave this value unmodified.
+            /// \return True if integration was successful, false otherwise. In that case, (q, v, a) are not updated.
+            virtual bool try_step(float64_t const& t,
```

Comment:
rather `(q, v, a, t, dt)` for "consistency with boost ?

Reply:
Yes and I don't know why. Functions are always written as `f(t, x)`, yet for some reasons numerical schemes are `scheme(f, x, t)``` - this is the case for boost and scipy for instance. I think it's because `t` is optional for autonomous dynamics (but then why isn't the same convention used for `f`.

Here I liked `f(t, x, dt)` better also because t is constant, while all following arguments are modified as reference: it fells weird to have a const in the middle of return values...

Reply:
No problem !